drv/i2c-devices/src/emc2305.rs

```diff
@@ -0,0 +1,334 @@
+    Fan1MinDrive = 0x38,
+    Fan1ValidTach = 0x39,
+    Fan1DriveFailBandLo = 0x3a,
+    Fan1DriveFailBandHi = 0x3b,
+    Tach1TargetLo = 0x3c,
+    Tach1TargetHi = 0x3d,
+    Tach1ReadingHi = 0x3e,
+    Tach1ReadingLo = 0x3f,
+
+    Fan2Setting = 0x40,
+    Pwm2Divide = 0x41,
+    Fan2Cfg1 = 0x42,
+    Fan2Cfg2 = 0x43,
+    Gain2 = 0x45,
+    Fan2SpinUpCfg = 0x46,
+    Fan2MaxStep = 0x47,
+    Fan2MinDrive = 0x48,
+    Fan2ValidTach = 0x49,
+    Fan2DriveFailBandLo = 0x4a,
+    Fan2DriveFailBandHi = 0x4b,
+    Tach2TargetLo = 0x4c,
+    Tach2TargetHi = 0x4d,
+    Tach2ReadingHi = 0x4e,
+    Tach2ReadingLo = 0x4f,
+
+    Fan3Setting = 0x50,
+    Pwm3Divide = 0x51,
+    Fan3Cfg1 = 0x52,
+    Fan3Cfg2 = 0x53,
+    Gain3 = 0x55,
+    Fan3SpinUpCfg = 0x56,
+    Fan3MaxStep = 0x57,
+    Fan3MinDrive = 0x58,
+    Fan3ValidTach = 0x59,
+    Fan3DriveFailBandLo = 0x5a,
+    Fan3DriveFailBandHi = 0x5b,
+    Tach3TargetLo = 0x5c,
+    Tach3TargetHi = 0x5d,
+    Tach3ReadingHi = 0x5e,
+    Tach3ReadingLo = 0x5f,
+
+    Fan4Setting = 0x60,
+    Pwm4Divide = 0x61,
+    Fan4Cfg1 = 0x62,
+    Fan4Cfg2 = 0x63,
+    Gain4 = 0x65,
+    Fan4SpinUpCfg = 0x66,
+    Fan4MaxStep = 0x67,
+    Fan4MinDrive = 0x68,
+    Fan4ValidTach = 0x69,
+    Fan4DriveFailBandLo = 0x6a,
+    Fan4DriveFailBandHi = 0x6b,
+    Tach4TargetLo = 0x6c,
+    Tach4TargetHi = 0x6d,
+    Tach4ReadingHi = 0x6e,
+    Tach4ReadingLo = 0x6f,
+
+    Fan5Setting = 0x70,
+    Pwm5Divide = 0x71,
+    Fan5Cfg1 = 0x72,
+    Fan5Cfg2 = 0x73,
+    Gain5 = 0x75,
+    Fan5SpinUpCfg = 0x76,
+    Fan5MaxStep = 0x77,
+    Fan5MinDrive = 0x78,
+    Fan5ValidTach = 0x79,
+    Fan5DriveFailBandLo = 0x7a,
+    Fan5DriveFailBandHi = 0x7b,
+    Tach5TargetLo = 0x7c,
+    Tach5TargetHi = 0x7d,
+    Tach5ReadingHi = 0x7e,
+    Tach5ReadingLo = 0x7f,
+
+    SoftwareLock = 0xef,
+    ProductFeatures = 0xfc,
+    ProductId = 0xfd,
+    MfgId = 0xfe,
+    Revision = 0xff,
+}
+
+pub struct Emc2305 {
+    pub device: I2cDevice,
+}
+
+pub const MAX_FANS: u8 = 5;
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct Fan(u8);
+
+impl TryFrom<u8> for Fan {
+    type Error = ();
+    /// Fans are based on a 0-based index. This should *not* be the number
+    /// of the fan (the fan numbers have a 1-based index)
+    fn try_from(index: u8) -> Result<Self, Self::Error> {
+        if index >= MAX_FANS {
+            Err(())
+        } else {
+            Ok(Self(index))
+        }
+    }
+}
+
+impl Fan {
+    fn register(&self, base: Register) -> Register {
+        Register::from_u8((base as u8) + self.0 * 0x10).unwrap()
+    }
+
+    fn configuration1(&self) -> Register {
+        self.register(Register::Fan1Cfg1)
+    }
+
+    fn configuration2(&self) -> Register {
+        self.register(Register::Fan1Cfg2)
+    }
+
+    fn tach_count(&self) -> Register {
+        self.register(Register::Tach1ReadingHi)
+    }
+
+    fn pwm_target(&self) -> Register {
+        self.register(Register::Fan1Setting)
+    }
```

Comment:
this is cute! part of my diseased brain kinda wonders if it could be made `const fn`, but i guess there wouldn't really be any benefit from that because we're looping over the fans at runtime anyway...

Reply:
Alas, the `from_u8` function from `#[derive(FromPrimitive)]` is not `const`.  I trust that the compiler will probably inline these functions into a few instructions, though.